packages/react-reconciler/src/ReactFiberHotReloading.js

```diff
@@ -0,0 +1,258 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {ReactElement} from 'shared/ReactElementType';
+import type {Fiber} from './ReactFiber';
+import type {FiberRoot} from './ReactFiberRoot';
+
+import {
+  flushSync,
+  scheduleWork,
+  flushPassiveEffects,
+} from './ReactFiberScheduler';
+import {Sync} from './ReactFiberExpirationTime';
+import {
+  FunctionComponent,
+  ForwardRef,
+  MemoComponent,
+  SimpleMemoComponent,
+} from 'shared/ReactWorkTags';
+import {REACT_FORWARD_REF_TYPE, REACT_MEMO_TYPE} from 'shared/ReactSymbols';
+
+type Family = {|
+  currentType: any,
+  currentSignature: null | string,
+|};
+
+type HotUpdate = {|
+  familiesByType: WeakMap<any, Family>,
+  root: FiberRoot,
+  staleFamilies: Set<Family>,
+  updatedFamilies: Set<Family>,
+|};
+
+let familiesByType: WeakMap<any, Family> | null = null;
+// $FlowFixMe Flow gets confused by a WeakSet feature check below.
+let failedBoundaries: WeakSet<Fiber> | null = null;
+
+export function resolveFunctionForHotReloading(type: any): any {
+  if (__DEV__) {
+    if (familiesByType === null) {
+      // Hot reloading is disabled.
+      return type;
+    }
+    let family = familiesByType.get(type);
+    if (family === undefined) {
+      return type;
+    }
+    // Use the latest known implementation.
+    return family.currentType;
+  } else {
+    return type;
+  }
+}
+
+export function resolveForwardRefForHotReloading(type: any): any {
+  if (__DEV__) {
+    if (familiesByType === null) {
+      // Hot reloading is disabled.
+      return type;
+    }
+    let family = familiesByType.get(type);
+    if (family === undefined) {
+      // Check if we're dealing with a real forwardRef. Don't want to crash early.
+      if (
+        type !== null &&
+        type !== undefined &&
+        typeof type.render === 'function'
+      ) {
+        // ForwardRef is special because its resolved .type is an object,
+        // but it's possible that we only have its inner render function in the map.
+        // If that inner render function is different, we'll build a new forwardRef type.
+        const currentRender = resolveFunctionForHotReloading(type.render);
+        if (type.render !== currentRender) {
+          return {...type, render: currentRender};
```

Comment:
Nit: No spread. You can use the Flow type in `ReactTypes` to make sure it stays in sync.

Reply:
I didn't want to lose `displayName` that might be declared on outer type. But I guess I can copy it over if it exists.

Reply:
Removed spread